src/librustc_trans/common.rs

```diff
@@ -347,65 +348,60 @@ impl<'a, 'tcx> FunctionContext<'a, 'tcx> {
             debuginfo::empty_function_debug_context(ccx)
         };
 
-        FunctionContext {
+        let mut fcx = FunctionContext {
             mir: mir,
             llfn: llfndecl,
-            llretslotptr: Cell::new(None),
+            llretslotptr: None,
             param_env: ccx.tcx().empty_parameter_environment(),
-            alloca_insert_pt: Cell::new(None),
+            alloca_insert_pt: None,
             landingpad_alloca: Cell::new(None),
             fn_ty: fn_ty,
             param_substs: param_substs,
             ccx: ccx,
             debug_context: debug_context,
             alloca_builder: OwnedBuilder::new_with_ccx(ccx),
-        }
-    }
+        };
 
-    /// Performs setup on a newly created function, creating the entry
-    /// scope block and allocating space for the return pointer.
-    pub fn init(&'a self, skip_retptr: bool) -> BlockAndBuilder<'a, 'tcx> {
-        let entry_bcx = self.build_new_block("entry-block");
+        let val = {
+            let entry_bcx = fcx.build_new_block("entry-block");
+            let val = entry_bcx.load(C_null(Type::i8p(ccx)));
+            fcx.alloca_builder.builder.position_at_start(entry_bcx.llbb());
+            val
+        };
 
         // Use a dummy instruction as the insertion point for all allocas.
-        // This is later removed in FunctionContext::cleanup.
-        self.alloca_insert_pt.set(Some(unsafe {
-            entry_bcx.load(C_null(Type::i8p(self.ccx)));
-            llvm::LLVMGetFirstInstruction(entry_bcx.llbb())
-        }));
-
-        self.alloca_builder.builder.position_at_start(entry_bcx.llbb());
-
-        if !self.fn_ty.ret.is_ignore() && !skip_retptr {
-            // We normally allocate the llretslotptr, unless we
-            // have been instructed to skip it for immediate return
-            // values, or there is nothing to return at all.
+        // This is later removed in the drop of FunctionContext.
+        fcx.alloca_insert_pt = Some(val);
```

Comment:
Can you avoid having two variables named `val`, i.e. can you move this up a bit more? Maybe rename it to `dummy`.